providers/enchant_nuspell.cpp

```diff
@@ -0,0 +1,444 @@
+	if (aff.size() >= 4 && aff.compare(aff.size() - 4, 4, ".dic") == 0) {
+		aff.erase(aff.size() - 3);
+		aff += "aff";
+	}
+	return aff;
+}
+
+static bool
+s_fileExists(const string & file)
+{
+	return g_file_test(file.c_str(), G_FILE_TEST_EXISTS) != 0;
+}
+
+static bool is_plausible_dict_for_tag(const char *dir_entry, const char *tag)
+{
+	const char *dic_suffix = ".dic";
+	size_t dic_suffix_len = strlen(dic_suffix);
+	size_t dir_entry_len = strlen(dir_entry);
+	size_t tag_len = strlen(tag);
+
+	if (dir_entry_len - dic_suffix_len < tag_len)
+		return false;
+	if (strcmp(dir_entry+dir_entry_len-dic_suffix_len, dic_suffix) != 0)
+		return false;
+	if (strncmp (dir_entry, tag, tag_len) != 0)
+		return false;
+	//e.g. requested dict for "fi",
+	//reject "fil_PH.dic"
+	//allow "fi-FOO.dic", "fi_FOO.dic", "fi.dic", etc.
+	if (!ispunct(dir_entry[tag_len]))
+		return false;
+	return true;
+}
+
+static char *
+nuspell_request_dictionary (const char * tag)
+{
+	vector<string> names;
+
+	s_buildHashNames (names, tag);
+
+	for (size_t i = 0; i < names.size (); i++) {
+		if (g_file_test(names[i].c_str(), G_FILE_TEST_EXISTS) &&
+		    s_fileExists(s_correspondingAffFile(names[i]))) {
+			return strdup (names[i].c_str());
+		}
+	}
+
+	vector<string> dirs;
+	s_buildDictionaryDirs (dirs);
+
+	for (size_t i = 0; i < dirs.size(); i++) {
+		GDir *dir = g_dir_open (dirs[i].c_str(), 0, nullptr);
+		if (dir) {
+			const char *dir_entry;
+			while ((dir_entry = g_dir_read_name (dir)) != NULL) {
+				if (is_plausible_dict_for_tag(dir_entry, tag)) {
+					char *dict = g_build_filename (dirs[i].c_str(),
+								       dir_entry, nullptr);
+					if(s_fileExists(s_correspondingAffFile(dict))) {
+						g_dir_close (dir);
+						return dict;
+					}
+					g_free(dict);
+				}
+			}
+
+			g_dir_close (dir);
+		}
+	}
+
+	return NULL;
+}
+
+bool
+NuspellChecker::requestDictionary(const char *szLang)
+{
+	char *dic = nuspell_request_dictionary (szLang);
+	if (!dic)
+		return false;
+	string aff(s_correspondingAffFile(dic));
+	if (!s_fileExists(aff))
+		return false;
+	auto path = string(dic);
+	if (path.size() >= 4 && path.compare(path.size() - 4, 4, ".dic") == 0)
+		path.erase(path.size() - 4);
+	else
+		return false;
+	try {
+		nuspell = nuspell::Dictionary::load_from_path(path);
+	} catch (const std::runtime_error& e) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Enchant
+ */
+
+static char **
+nuspell_dict_suggest (EnchantDict * me, const char *const word,
+		      size_t len, size_t * out_n_suggs)
+{
+	NuspellChecker * checker = static_cast<NuspellChecker *>(me->user_data);
+	return checker->suggestWord (word, len, out_n_suggs);
+}
+
+static int
+nuspell_dict_check (EnchantDict * me, const char *const word, size_t len)
+{
+	NuspellChecker * checker = static_cast<NuspellChecker *>(me->user_data);
+
+	if (checker->checkWord(word, len))
+		return 0;
+
+	return 1;
+}
+
+static int
+nuspell_dict_is_word_character (EnchantDict *me, uint32_t uc, size_t n)
+{
+	(void)n;
```

Comment:
Instead of casting `n` to `void`, use `_GL_UNUSED_PARAMETER`.

Reply:
(Also needed for `me`.)

Reply:
Results in error:

    enchant_nuspell.cpp: In function ‘int nuspell_dict_is_word_character(EnchantDict*, uint32_t, size_t)’:
    enchant_nuspell.cpp:293:3: error: expected primary-expression before ‘__attribute__’
      293 |  (_GL_UNUSED_PARAMETER)n;
          |   ^~~~~~~~~~~~~~~~~~~~
    enchant_nuspell.cpp:293:3: error: expected ‘)’ before ‘__attribute__’
      293 |  (_GL_UNUSED_PARAMETER)n;
          |  ~^
          |   )

Reply:
Does this suffice instead? https://github.com/nuspell/enchant/commit/f9a7ff72210cac77efdf1a844379f554299a0fe7#diff-b4a4b7f548e6522299f21f4f5632ff1cR293

Reply:
Fixed in https://github.com/nuspell/enchant/commit/d8315276e58c13e701fe06456eda11e633d36d06#diff-b4a4b7f548e6522299f21f4f5632ff1cR291